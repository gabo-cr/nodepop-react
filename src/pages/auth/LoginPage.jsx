import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";

import { getUi } from "../../store/selectors";
import { authLogin, uiResetError } from "../../store/actions";

import { FormInput } from "../../components/shared/form/input/FormInput";
import Alert from "../../components/shared/alert/Alert";
import AuthLayout from "../../components/layout/authLayout/AuthLayout";
import Button from "../../components/shared/button/Button";

export default function LoginPage() {
	//** Hooks */
	const dispatch = useDispatch();
	const { pending: isFetching, error } = useSelector(getUi);
	
	//** States */
	const [formValues, setFormValues] = useState({
		email: '',
		password: '',
	});
	const [rememberme, setRememberme] = useState(false);

	//** Handlers */
	const handleChange = (event) => {
		setFormValues(currentFormValues => ({
			...currentFormValues,
			[event.target.name]: event.target.value,
		}));
	};

	const handleChangeRememberme = (event) => {
		setRememberme(event.target.checked);
	};

	const handleSubmit = async (event) => {
		event.preventDefault();
		dispatch(authLogin(formValues, rememberme));
	};

	const resetError = () => dispatch(uiResetError());

	const { email, password } = formValues;
	const buttonDisabled = !email || !password || isFetching;

	return (
		<AuthLayout title="Iniciar sesión">
			<div className="loginPage">
				<form onSubmit={handleSubmit}>
					<FormInput label="Email" name="email" aria-label="email" id="email" type="text" value={email} onChange={handleChange} />
					<FormInput label="Password" name="password" aria-label="password" id="password" type="password" value={password} onChange={handleChange} />
					<FormInput label="Recuérdame" name="rememberme" aria-label="rememberme" id="rememberme" type="checkbox" checked={rememberme} onChange={handleChangeRememberme} />
					<Button type="submit" disabled={buttonDisabled} fullwidth>Log in</Button> 
				</form>
				{error && (
					<Alert variant="error">
						<div onClick={resetError}>
							{error.message}
						</div>
					</Alert>
				)}
			</div>
		</AuthLayout>
	);
}