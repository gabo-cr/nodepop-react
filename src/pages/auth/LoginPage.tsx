import { ChangeEvent, SyntheticEvent, useState } from "react";
import { FormInput } from "../../components/shared/FormInput";
import { login } from "../../api/auth";
import { useLocation, useNavigate } from "react-router-dom";
import { LoginParameters, LoginResponseError } from "../../types/auth";
import { useAuth } from "../../context/AuthContextProvider";

export default function LoginPage() {
	//** Hooks */
	const location = useLocation();
  	const navigate = useNavigate();
	const { onLogin } = useAuth();

	//** States */
	const [isFetching, setIsFetching] = useState<boolean>(false);
	const [formValues, setFormValues] = useState<LoginParameters>({
		email: '',
		password: '',
	});
	const [rememberme, setRememberme] = useState<boolean>(false);
	const [error, setError] = useState<LoginResponseError | null>(null);

	//** Handlers */
	const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
		setFormValues(currentFormValues => ({
			...currentFormValues,
			[event.target.name]: event.target.value,
		}));
	};

	const handleChangeRememberme = (event: ChangeEvent<HTMLInputElement>) => {
		setRememberme(event.target.checked);
	};

	const handleSubmit = async (event: SyntheticEvent) => {
		event.preventDefault();

		try {
			setIsFetching(true);
			await login(formValues, rememberme);
			setIsFetching(false);
			onLogin();
			const to = location.state?.from || '/';
			navigate(to, { replace: true });
		} catch (error: any) {
			setIsFetching(false);
			setError(error);
		}
	};

	const resetError = () => setError(null);

	const { email, password } = formValues;
	const buttonDisabled = !email || !password || isFetching;

	return (
		<div className="loginPage">
			<h1>Iniciar sesi√≥n</h1>
			<form onSubmit={handleSubmit}>
				<FormInput label="Email" name="email" id="email" type="text" value={email} onChange={handleChange} />
				<FormInput label="Password" name="password" id="password" type="password" value={password} onChange={handleChange} />
				<input type="checkbox" name="rememberme" id="rememberme" checked={rememberme} onChange={handleChangeRememberme} />
				<button type="submit" disabled={buttonDisabled}>Log in</button> 
			</form>
			{error && (
				<div className="loginPage-error" onClick={resetError}>
					{error.message}
				</div>
			)}
		</div>
	);
}