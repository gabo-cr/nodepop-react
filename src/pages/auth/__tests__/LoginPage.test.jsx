import React from 'react';
import { render } from '@testing-library/react';
import LoginPage from '../LoginPage';
import { Provider } from 'react-redux';
import { RouterProvider, createBrowserRouter } from 'react-router-dom';
import configureStore from '../../../store';

describe('LoginPage', () => {
	const router = createBrowserRouter([{ path: '*', element: <LoginPage /> }]);
	const store = configureStore({
		auth: false,
		adverts: {
			data: [],
			loaded: false
		},
		tags: {
			data: [],
			loaded: false
		},
		ui: {
			pending: false,
			error: undefined
		}
	}, { router });
	
	const renderComponent = () => render(
		<Provider store={store}>
			<RouterProvider router={router} />
		</Provider>
	);

	test('snapshot', () => {
		const { container } = renderComponent();
		expect(container).toMatchSnapshot();
	});
});
