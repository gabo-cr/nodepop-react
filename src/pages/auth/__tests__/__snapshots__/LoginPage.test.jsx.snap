// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="authLayout"
  >
    <main>
      <h1>
        Iniciar sesión
      </h1>
      <div
        class="loginPage"
      >
        <form>
          <div
            class="formField"
          >
            <label
              class="label"
            >
              Email
            </label>
            <input
              aria-label="email"
              autocomplete="off"
              class="input"
              id="email"
              name="email"
              type="text"
              value=""
            />
          </div>
          <div
            class="formField"
          >
            <label
              class="label"
            >
              Password
            </label>
            <input
              aria-label="password"
              autocomplete="off"
              class="input"
              id="password"
              name="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="formCheck"
          >
            <input
              aria-label="rememberme"
              id="rememberme"
              name="rememberme"
              type="checkbox"
            />
            <span>
              Recuérdame
            </span>
          </div>
          <button
            class="btn primary"
            disabled=""
            style="width: 100%;"
            type="submit"
          >
            Log in
          </button>
        </form>
      </div>
    </main>
  </div>
</div>
`;
